Preliminary support for unions added.

harpocrates opened the pull request:
This is really _far_ from complete, but it starts addressing #22 by filling in everything straightforward. Here are the issues AFAICT
- We cannot yet derive `Copy` and `Clone` for unions, but support for that [is pending](https://github.com/rust-lang/rust/pull/36384) (incredibly fast given how recent an addition unions are!).
- We'll need to add `#![feature(untagged_unions)]` to the top of files (I don't think we currently have anything to do this - if we don't it'd be worth it to make it good).
- The can of worms that was struct initialization needs to be reopened. I'll take a look at the C99 spec. I'm _hoping_ that when there are multiple (named) initializers in the initializer list for a union in C99, we simply ignore all except the last one (even if the previous ones are bigger), and at least [clang](https://gcc.godbolt.org/#compilers:!%28%28compiler:clang380,options:%27-O2+-Wall+-xc%27,sourcez:MQSwdgxgNgrgJgUwAQB4DOAXOID2A6ACwD4AoEzAJxggyQBUkBvEpJcWjARgG4W2wOAJl4BfXpWq0Aykz4Sa9JGh4kxZdkgC2AQ3AAKAG44QcAJQlmrebQa0AvEyR4MgpA9dqrGKgplo3sqyseMoBGAA0fMHKzpwBnJGsnkgADhTsAGZ6AEQApHBI%2BQA6YNnhSiGcseVolc6CpqJAAA%3D%29%29,filterAsm:%28commentOnly:!t,directives:!t,labels:!t%29,version:3) seems to agree... I'm also hoping that the rust unions are also initialized with an extra empty space zeroed out. If all that works out as expected, I'll still need to find a nice way of expressing union based initializers - something like
  
  ``` haskell
  data Initializer
     = StructInitializer (Maybe Rust.Expr) (IntMap.IntMap Initializer)
     = UnionInitializer Int Initializer
  ```

Is there anything else I'm missing here? I think there is something to do with incomplete types that you did for structs that also applies to unions. I'm thinking of [this](https://github.com/jameysharp/corrode/commit/3346deb4005e12bc27e40f63132b62953859e9e3)...

harpocrates committed:
Preliminary support for unions added.

petrochenkov commented:
> We cannot yet derive Copy and Clone for unions

FYI, `Copy` can already be derived, it was a part of initial union implementation.

harpocrates committed:
Adds support for initialization of unions

Apart from a couple of obvious bug fixes, this focuses on initialization
for unions. The previous initialization code had been designed with only struct
and arrays in mind, so some pretty deep changes had to be made.

   * the data type `Initializer` now has three constructors for regular scalar
     initializers, aggregate initializers (arrays and structs), and union
     initializers. Might want to think about naming `Scalar` differently (see
     section 6.2.5 point 21).
   * `Initializer` is no longer a `Monoid`, since we can't make a `mempty`
     without knowing the type we are initializing. Instead, it is a `Semigroup`,
     with some combinations not valid. This means that either we will need
     to enforce `base >= 4.9` or `semigroups`.

To the best of my current knowledge and testing, this accurately and completely
implements initialization of unions as per C99.

harpocrates commented:
This build fails because one of the big changes that comes with this last commit (that adds what I think is complete support for union initialization) is that `Initializer` no longer is a `Monoid` but a `Semigroup` instead. That means we'll need to add either `semigroups` or `base >= 4.9` to the cabal file, and currently I've done neither.

Personally I am in favour of the latter (if we are generating rust that only compiles on rust nightly, it seems fairly reasonably to say we run on GHC 8.0), but I'll leave the choice up to you.

harpocrates committed:
Use `IsIncomplete` for union types too

Add support for incomplete unions in the context of:

  * completing them
  * performing member access on them
  * zero initializing them

Also added missing case in `Eq CType` for unions.

zaoqi commented:
@harpocrates Conflicting files
